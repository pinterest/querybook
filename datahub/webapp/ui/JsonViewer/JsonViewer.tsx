import React from 'react';
import styled from 'styled-components';

const StyledJsonViewer = styled.div`
    background-color: var(--light-bg-color);
    color: var(--dark-text-color);
    font-size: var(--small-text-size);
    overflow-x: auto;
    padding: 20px 24px;
    white-space: pre;
    word-wrap: normal;
`;

export interface IJsonViewerProps {
    value: string | {};
    indent?: number;
    replacerFunction?: () => any;
}

export const JsonViewer: React.FunctionComponent<IJsonViewerProps> = ({
    value = '',
    replacerFunction: defaultReplacer,
    indent = 4,
}) => {
    let jsonString = '';
    if (typeof value === 'object') {
        const replacerFunction = defaultReplacer
            ? defaultReplacer.bind(null, value)
            : null;

        jsonString = JSON.stringify(value, replacerFunction, indent);
    } else if (typeof value === 'string') {
        try {
            const valueObj = JSON.parse(value);
            const replacerFunction = defaultReplacer
                ? defaultReplacer.bind(null, valueObj)
                : null;
            jsonString = JSON.stringify(valueObj, replacerFunction, indent);
        } catch (err) {
            jsonString = `Could not parse ${value} as JSON. Please check the syntax.`;
        }
    } else {
        jsonString = '"value" is an unexpected type of: ' + typeof value;
    }

    return (
        <StyledJsonViewer className="JsonViewer">{jsonString}</StyledJsonViewer>
    );
};
