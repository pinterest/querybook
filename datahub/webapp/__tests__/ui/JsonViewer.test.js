import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import renderer from 'react-test-renderer';

import { JsonViewer } from '../../ui/JsonViewer/JsonViewer';

it('renders without crashing', () => {
    shallow(<JsonViewer value={{ test: 'test' }} />);
});

describe('matches enzyme snapshots', () => {
    it('matches snapshot', () => {
        let wrapper = shallow(<JsonViewer value={{ test: 'test' }} />);
        let serialized = toJson(wrapper);
        expect(serialized).toMatchSnapshot();
    });
});

describe('matches test renderer snapshot', () => {
    it('serializes the styles', () => {
        const output = renderer.create(<JsonViewer value={{ test: 'test' }} />);
        expect(output).toMatchSnapshot();
    });
});
