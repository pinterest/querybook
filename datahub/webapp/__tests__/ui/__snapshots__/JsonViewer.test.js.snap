// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches enzyme snapshots matches snapshot 1`] = `
<styled.div
  className="JsonViewer"
>
  {
    "test": "test"
}
</styled.div>
`;

exports[`matches test renderer snapshot serializes the styles 1`] = `
<div
  className="JsonViewer sc-bdVaJa cyihmS"
>
  {
    "test": "test"
}
</div>
`;
